src/mixins/themeable.ts

```diff
@@ -43,78 +51,74 @@ export interface ThemeableOptions {
+const negativeClassMap = new WeakMap<Themeable<any>, FlaggedCSSModuleClassNames>();
 
-function addClassNameToCSSModuleClassNames(cssModuleClassNames: CSSModuleClassNames, classList: StringIndexedObject, className: string) {
-	if (classList.hasOwnProperty(className)) {
-		// split out the classname because css-module composition combines class names with a space
-		const generatedClassNames: string[] = classList[className].split(' ');
-		generatedClassNames.forEach((generatedClassName) => {
-			cssModuleClassNames[generatedClassName] = true;
-		});
-	}
+function updateNegativeClassNames<T>(instance: Themeable<T>, classNames: string[]) {
+	const negativeClassFlags = setClassNameFlags(classNames, false);
+	const currentNegativeClassFlags = negativeClassMap.get(instance);
+	negativeClassMap.set(instance, assign({}, currentNegativeClassFlags, negativeClassFlags));
 }
 
-function negatePreviousClasses<T>(previousClasses: AppliedClasses<T>, newClasses: AppliedClasses<T>) {
-	return Object.keys(previousClasses).reduce((newAppliedClasses, className: keyof T) => {
-		const oldCSSModuleClassNames = <CSSModuleClassNames> previousClasses[className];
+function setClassNameFlags(classNames: string[], applied: boolean) {
+	return classNames.reduce((flaggedClassNames: FlaggedCSSModuleClassNames, className) => {
+		flaggedClassNames[className] = applied;
+		return flaggedClassNames;
+	}, <FlaggedCSSModuleClassNames> {});
+}
+
+function generateThemeClasses<T>(instance: Themeable<T>, { classes: baseThemeClasses, path }: BaseTheme<T>, theme: any = {}, overrideClasses: any = {}) {
+	const applicableThemeClasses = theme.hasOwnProperty(path) ? theme[path] : {};
+	const combinedThemeClasses = assign({}, baseThemeClasses, applicableThemeClasses);
+	let allClasses: string[] = [];
 
-		const negatedCSSModuleClassNames = Object.keys(oldCSSModuleClassNames).reduce((newCSSModuleClassNames, oldCSSModuleClassName) => {
-			const currentClassNameFlag = oldCSSModuleClassNames[oldCSSModuleClassName];
-			// If it's true it needs to be negated and passed along, If it's false,
-			// don't return it as maquette will already have removed it.
-			if (currentClassNameFlag) {
-				newCSSModuleClassNames[oldCSSModuleClassName] = false;
-			}
-			return newCSSModuleClassNames;
-		}, <CSSModuleClassNames> {});
+	const themeClasses = Object.keys(combinedThemeClasses).reduce((newAppliedClasses, className: keyof T) => {
+		const cssClassNames = combinedThemeClasses[className].split(' ');
+		const overrideCssClassNames = overrideClasses.hasOwnProperty(className) ? overrideClasses[className].split(' ') : [];
+		const combinedCssClassNames = [...cssClassNames, ...overrideCssClassNames];
+		allClasses = [...allClasses, ...combinedCssClassNames];
 
-		const calculatedClassNameMap = assign({}, negatedCSSModuleClassNames, newClasses[className]);
-		newAppliedClasses[className] = calculatedClassNameMap;
+		newAppliedClasses[className] = setClassNameFlags(combinedCssClassNames, true);
```

Comment:
Given that we have Flagged, Override, Combined, Applied, New, Negative, etc., I think we need somewhere a summary of what all of these terms mean, perhaps at the top of the file?